ocaml/xapi/xapi_ha_vm_failover.ml

```diff
@@ -580,52 +731,80 @@ let restart_auto_run_vms ~__context live_set n =
+             | Some h -> Client.Client.Async.VM.start_on rpc session_id vm h false true
            end else failwith (Printf.sprintf "VM: %s restart attempt delayed for 120s" (Ref.string_of vm)) in
-         try
-           go ();
-           true
-         with
-         | Api_errors.Server_error(code, params) when code = Api_errors.ha_operation_would_break_failover_plan ->
+         TaskChains.task go >>= function
+         | Result.Error Api_errors.Server_error(code, params) when code = Api_errors.ha_operation_would_break_failover_plan ->
            (* This should never happen since the planning code would always allow the restart of a protected VM... *)
            error "Caught exception HA_OPERATION_WOULD_BREAK_FAILOVER_PLAN: setting pool as overcommitted and retrying";
            ignore (mark_pool_as_overcommitted ~__context ~live_set : bool);
            begin
-             try
-               go ();
-               true
-             with e ->
-               error "Caught exception trying to restart VM %s: %s" (Ref.string_of vm) (ExnHelper.string_of_exn e);
-               false
+             TaskChains.task go >>= function
+             | Result.Ok _ -> TaskChains.ok ()
+             | Result.Error e -> error "Caught exception trying to restart VM %s: %s" (Ref.string_of vm) (ExnHelper.string_of_exn e);
+               TaskChains.fail e
            end
-         | e ->
+         | Result.Error e ->
            error "Caught exception trying to restart VM %s: %s" (Ref.string_of vm) (ExnHelper.string_of_exn e);
-           false in
+           TaskChains.fail e
+         | Result.Ok _ -> TaskChains.ok ()
+       in
+
+       (** [ordered_map_concat f lst]
+        * traverses the list in order, maps each inner list through [f]
+        * and concatenates the result *)
+       let ordered_map_concat f lst =
+        debug "Processing %d parallel groups" (List.length lst);
+        (* the iteration order is important here for preserving the VM start order *)
+        List.fold_left (fun accum inner ->
+             (* the order of elements in the result doesn't matter,
+              * they were launched in parallel *)
+             List.rev_append (f inner) accum) [] lst
+       in
+
+       (** [map_parallel ~order_f f lst] groups objects in [lst] by order number 
+           (provided by applying [order_f] to each object). For each group of objects
+           it then applies [f] to them to produce an 'a TaskChain.t and then executes this
+           set of TaskChain.t values in parallel until they each of the TaskChain.t 
+           values has been executed to completion. It then begins work on the next group
+           of objects. 
+
+           The return is a list of 'a Result.results
+       *)
+       let map_parallel ~order_f f lst =
+         lst
+         |> Helpers.group_by ~ordering:`ascending order_f
+         |> ordered_map_concat (fun same_order ->
+             same_order
+             |> List.rev_map fst
```

Comment:
To make it slightly more efficient, `map` and `fst` could be combined into `List.map (fun xs -> f (fst xs))`